File: frontend/src/api/log.ts
import { useQuery, useQueryClient } from "react-query";
import { Modify } from "./dataTypes";
import { fetchJsonGet } from "../utils/fetch";

export function useLog(page = 0)
{
	const queryClient = useQueryClient();
	const fn = async (p: number) =>
		(await fetchJsonGet<LogItemResponse[]>(`/api/debug/log?p=${p}`)).map(parseLogResponse);
	queryClient.prefetchQuery(["log", page - 1], () => fn(page - 1));
	queryClient.prefetchQuery(["log", page + 1], () => fn(page + 1));
	return useQuery(["log", page], () => fn(page));
}

export function useLogCacheClear()
{
	const queryClient = useQueryClient();
	return () =>
	{
		queryClient.removeQueries("log");
		queryClient.invalidateQueries("logLen");
	}
}

export function useLogLen()
{
	return useQuery("logLen", async () =>
		await fetchJsonGet<LogLenResponse>("/api/debug/log_len")
	);
}

export function parseLogResponse(responseLog: LogItemResponse)
{
	const log = responseLog as unknown as LogItem;
	log.date = new Date(responseLog.date);
	return log;
}

export interface LogItem
{
	id: number,
	actionCode: "added" | "updated" | "deleted" | "restored",
	date: Date,
	recordId: number,
	userId: number,
	userName: string,
	tableName: "User" | "Role" | "Event" | "Ticket" | "TicketType" | "PermissionAccess",
	changes: [string, any, any][],
}

type LogItemResponse = Modify<LogItem, {
	date: string,
}>

export interface LogLenResponse
{
	len: number,
}

File: backend/data/log.py
from sqlalchemy import Column, DateTime, ForeignKey, orm, Integer, String, JSON
from sqlalchemy_serializer import SerializerMixin
from .db_session import SqlAlchemyBase


class Log(SqlAlchemyBase, SerializerMixin):
    __tablename__ = "Log"

    id         = Column(Integer, primary_key=True, autoincrement=True, unique=True)
    date       = Column(DateTime, nullable=False)
    actionCode = Column(String(16), nullable=False)
    userId     = Column(Integer, ForeignKey("User.id"), nullable=False)
    userName   = Column(String(64), nullable=False)
    tableName  = Column(String(16), nullable=False)
    recordId   = Column(Integer, nullable=False)
    changes    = Column(JSON, nullable=False)

    user = orm.relationship("User")

    def __repr__(self):
        return f"<Log> [{self.id}] {self.date} {self.actionCode}"

    def get_dict(self):
        return self.to_dict(only=("id", "date", "actionCode", "userId", "userName", "tableName", "recordId", "changes"))


class Actions:
    added = "added"
    updated = "updated"
    deleted = "deleted"
    restored = "restored"


class Tables:
    User = "User"
    Role = "Role"
    Event = "Event"
    Ticket = "Ticket"
    TicketType = "TicketType"
    PermissionAccess = "PermissionAccess"
    Image = "Image"
